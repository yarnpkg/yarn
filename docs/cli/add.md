# `add`

## `kpm add <package...>`

## `kpm add <package...> --dev`

## `kpm add <package...> --peer`

## `kpm add <package...> --optional`

## `kpm add <package...> --exact`

## `kpm add <package...> --tilde`
