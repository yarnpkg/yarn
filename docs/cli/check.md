# `check`

## `kpm check`

## `kpm check --integrity`
