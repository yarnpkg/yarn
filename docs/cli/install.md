# `install`

## `kpm install`

## `kpm install --production`

## `kpm install --force`

## `kpm install --no-lockfile`
