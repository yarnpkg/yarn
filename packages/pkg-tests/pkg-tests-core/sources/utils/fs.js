/* @flow */

import type {Gzip} from 'zlib';

const fs = require('fs-extra');
const klaw = require('klaw');
const path = require('path');
const tarFs = require('tar-fs');
const tmp = require('tmp');
const zlib = require('zlib');

const miscUtils = require('./misc');

exports.walk = function walk(
  source: string,
  {filter, relative = false}: {|filter?: Array<string>, relative?: boolean|} = {},
): Promise<Array<string>> {
  return new Promise((resolve, reject) => {
    const paths = [];

    const walker = klaw(source, {
      filter: itemPath => {
        if (!filter) {
          return true;
        }

        const stat = fs.statSync(itemPath);

        if (stat.isDirectory()) {
          return true;
        }

        const relativePath = path.relative(source, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) {
          return true;
        }

        return false;
      },
    });

    walker.on('data', ({path: itemPath}) => {
      const relativePath = path.relative(source, itemPath);

      if (!filter || miscUtils.filePatternMatch(relativePath, filter)) {
        paths.push(relative ? relativePath : itemPath);
      }

      // This item has been accepted only because it's a directory; it doesn't match the filter
      return;
    });

    walker.on('end', () => {
      resolve(paths);
    });
  });
};

exports.packToStream = function packToStream(
  source: string,
  // $FlowFixMe Flow doesn't support null default parameters in exact types
  {virtualPath = null}: {|virtualPath?: ?string|} = {},
): Gzip {
  if (virtualPath) {
    if (!path.isAbsolute(virtualPath)) {
      throw new Error('The virtual path has to be an absolute path');
    } else {
      virtualPath = path.resolve(virtualPath);
    }
  }

  const zipperStream = zlib.createGzip();

  const packStream = tarFs.pack(source, {
    map: header => {
      if (true) {
        header.name = path.resolve('/', header.name);
        header.name = path.relative('/', header.name);
      }

      if (virtualPath) {
        header.name = path.resolve('/', virtualPath, header.name);
        header.name = path.relative('/', header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on('error', error => {
    zipperStream.emit('error', error);
  });

  return zipperStream;
};

exports.packToFile = function packToFile(target: string, source: string, options: *): Promise<void> {
  const tarballStream = fs.createWriteStream(target);

  const packStream = exports.packToStream(source, options);
  packStream.pipe(tarballStream);

  return new Promise((resolve, reject) => {
    tarballStream.on('error', error => {
      reject(error);
    });

    packStream.on('error', error => {
      reject(error);
    });

    tarballStream.on('close', () => {
      resolve();
    });
  });
};

exports.createTemporaryFolder = function createTemporaryFolder(): Promise<string> {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, dirPath) => {
      if (error) {
        reject(error);
      } else {
        resolve(dirPath);
      }
    });
  });
};

exports.createTemporaryFile = async function createTemporaryFile(filePath: string): Promise<string> {
  if (filePath) {
    if (path.normalize(filePath).match(/^(\.\.)?\//)) {
      throw new Error('A temporary file path must be a forward path');
    }

    const folderPath = await exports.createTemporaryFolder();
    return path.resolve(folderPath, filePath);
  } else {
    return new Promise((resolve, reject) => {
      tmp.file({discardDescriptor: true}, (error, filePath) => {
        if (error) {
          reject(error);
        } else {
          resolve(filePath);
        }
      });
    });
  }
};

exports.writeFile = async function writeFile(target: string, body: string | Buffer): Promise<void> {
  await fs.mkdirp(path.dirname(target));
  await fs.writeFile(target, body);
};

exports.readFile = function readFile(source: string, encoding: ?string = null): Promise<any> {
  return fs.readFile(source, encoding);
};

exports.writeJson = function writeJson(target: string, object: any): Promise<void> {
  return exports.writeFile(target, JSON.stringify(object));
};

exports.readJson = async function readJson(source: string): Promise<any> {
  const fileContent = await exports.readFile(source, 'utf8');

  try {
    return JSON.parse(fileContent);
  } catch (error) {
    throw new Error(`Invalid json file (${source})`);
  }
};
