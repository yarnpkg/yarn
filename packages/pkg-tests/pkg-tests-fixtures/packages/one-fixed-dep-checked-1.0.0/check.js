process.exit(require('no-deps-checked/built') === 42 ? 0 : 1);
