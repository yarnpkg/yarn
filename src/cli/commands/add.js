/* @flow */

import type {Reporter} from '../../reporters/index.js';
import type {InstallCwdRequest} from './install.js';
import type {DependencyRequestPatterns, Manifest} from '../../types.js';
import type Config from '../../config.js';
import type {ListOptions} from './list.js';
import Lockfile from '../../lockfile/wrapper.js';
import PackageRequest from '../../package-request.js';
import {buildTree} from './list.js';
import {wrapLifecycle, Install} from './install.js';
import {MessageError} from '../../errors.js';

const invariant = require('invariant');

export class Add extends Install {
  constructor(args: Array<string>, flags: Object, config: Config, reporter: Reporter, lockfile: Lockfile) {
    super(flags, config, reporter, lockfile);
    this.args = args;
    // only one flag is supported, so we can figure out which one was passed to `yarn add`
    this.flagToOrigin = [
      flags.dev && 'devDependencies',
      flags.optional && 'optionalDependencies',
      flags.peer && 'peerDependencies',
      'dependencies',
    ]
      .filter(Boolean)
      .shift();
  }

  args: Array<string>;
  flagToOrigin: string;
  addedPatterns: Array<string>;

  /**
   * TODO
   */

  prepareRequests(requests: DependencyRequestPatterns): DependencyRequestPatterns {
    const requestsWithArgs = requests.slice();

    for (const pattern of this.args) {
      requestsWithArgs.push({
        pattern,
        registry: 'npm',
        optional: false,
      });
    }
    return requestsWithArgs;
  }

  /**
   * returns version for a pattern based on Manifest
   */
  getPatternVersion(pattern: string, pkg: Manifest): string {
    const {exact, tilde} = this.flags;
    const parts = PackageRequest.normalizePattern(pattern);
    let version;
    if (PackageRequest.getExoticResolver(pattern)) {
      // wasn't a name/range tuple so this is just a raw exotic pattern
      version = pattern;
    } else if (parts.hasVersion && parts.range) {
      // if the user specified a range then use it verbatim
      version = parts.range === 'latest' ? `^${pkg.version}` : parts.range;
    } else if (tilde) {
      // --save-tilde
      version = `~${pkg.version}`;
    } else if (exact) {
      // --save-exact
      version = pkg.version;
    } else {
      // default to save prefix
      version = `${String(this.config.getOption('save-prefix') || '')}${pkg.version}`;
    }
    return version;
  }

  preparePatterns(patterns: Array<string>): Array<string> {
    const preparedPatterns = patterns.slice();
    for (const pattern of this.resolver.dedupePatterns(this.args)) {
      const pkg = this.resolver.getResolvedPattern(pattern);
      invariant(pkg, `missing package ${pattern}`);
      const version = this.getPatternVersion(pattern, pkg);
      const newPattern = `${pkg.name}@${version}`;
      preparedPatterns.push(newPattern);
      this.addedPatterns.push(newPattern);
      if (newPattern === pattern) {
        continue;
      }
      this.resolver.replacePattern(pattern, newPattern);
    }
    return preparedPatterns;
  }

  bailout(patterns: Array<string>): Promise<boolean> {
    return Promise.resolve(false);
  }

  /**
   * Description
   */

  async init(): Promise<Array<string>> {
    this.addedPatterns = [];
    const patterns = await Install.prototype.init.call(this);
    await this.maybeOutputSaveTree(patterns);
    await this.savePackages();
    return patterns;
  }

  /**
   * Description
   */

  fetchRequestFromCwd(): Promise<InstallCwdRequest> {
    return Install.prototype.fetchRequestFromCwd.call(this, this.args);
  }

  /**
   * Output a tree of any newly added dependencies.
   */

  async maybeOutputSaveTree(patterns: Array<string>): Promise<void> {
    // don't limit the shown tree depth
    const opts: ListOptions = {
      reqDepth: 0,
    };
    const {trees, count} = await buildTree(this.resolver, this.linker, patterns, opts, true, true);
    this.reporter.success(
      count === 1 ? this.reporter.lang('savedNewDependency') : this.reporter.lang('savedNewDependencies', count),
    );
    this.reporter.tree('newDependencies', trees);
  }

  /**
   * Save added packages to manifest if any of the --save flags were used.
   */

  async savePackages(): Promise<void> {
    // fill rootPatternsToOrigin without `excludePatterns`
    await Install.prototype.fetchRequestFromCwd.call(this);
    const patternOrigins = Object.keys(this.rootPatternsToOrigin);

    // get all the different registry manifests in this folder
    const manifests = await this.config.getRootManifests();

    // add new patterns to their appropriate registry manifest
    for (const pattern of this.addedPatterns) {
      const pkg = this.resolver.getResolvedPattern(pattern);
      invariant(pkg, `missing package ${pattern}`);
      const version = this.getPatternVersion(pattern, pkg);
      const ref = pkg._reference;
      invariant(ref, 'expected package reference');
      // lookup the package to determine dependency type; used during `yarn upgrade`
      const depType = patternOrigins.reduce((acc, prev) => {
        if (prev.indexOf(`${pkg.name}@`) === 0) {
          return this.rootPatternsToOrigin[prev];
        }
        return acc;
      }, null);

      // depType is calculated when `yarn upgrade` command is used
      const target = depType || this.flagToOrigin;

      // add it to manifest
      const {object} = manifests[ref.registry];

      object[target] = object[target] || {};
      object[target][pkg.name] = version;
    }

    await this.config.saveRootManifests(manifests);
  }
}

export function hasWrapper(commander: Object): boolean {
  return true;
}

export function setFlags(commander: Object) {
  commander.usage('add [packages ...] [flags]');
  commander.option('-D, --dev', 'save package to your `devDependencies`');
  commander.option('-P, --peer', 'save package to your `peerDependencies`');
  commander.option('-O, --optional', 'save package to your `optionalDependencies`');
  commander.option('-E, --exact', 'install exact version');
  commander.option('-T, --tilde', 'install most recent release with the same minor version');
}

export async function run(config: Config, reporter: Reporter, flags: Object, args: Array<string>): Promise<void> {
  if (!args.length) {
    throw new MessageError(reporter.lang('missingAddDependencies'));
  }

  const lockfile = await Lockfile.fromDirectory(config.lockfileFolder, reporter);

  await wrapLifecycle(config, flags, async () => {
    const install = new Add(args, flags, config, reporter, lockfile);
    await install.init();
  });
}
