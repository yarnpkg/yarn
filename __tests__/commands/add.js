/* @flow */

import {getPackageVersion, createLockfile, explodeLockfile, run as buildRun} from './_install.js';
import {Add} from '../../src/cli/commands/add.js';
import {Reporter} from '../../src/reporters/index.js';
import * as constants from '../../src/constants.js';
import {parse} from '../../src/lockfile/wrapper.js';
import {Install} from '../../src/cli/commands/install.js';
import Lockfile from '../../src/lockfile/wrapper.js';
import {run as check} from '../../src/cli/commands/check.js';
import Config from '../../src/config.js';
import * as fs from '../../src/util/fs.js';
import {runInstall} from './_install.js';
import assert from 'assert';
import semver from 'semver';

jasmine.DEFAULT_TIMEOUT_INTERVAL = 90000;

const path = require('path');

const fixturesLoc = path.join(__dirname, '..', 'fixtures', 'add');

function runAdd(
  flags: Object,
  args: Array<string>,
  name: string,
  checkInstalled?: ?(config: Config, reporter: Reporter) => ?Promise<void>,
  beforeInstall?: ?(cwd: string) => ?Promise<void>,
  cleanupAfterInstall: boolean = true,
): Promise<void> {
  return buildRun((config, reporter, lockfile): Install => {
    return new Add(args, flags, config, reporter, lockfile);
  }, flags, path.join(fixturesLoc, name), checkInstalled, beforeInstall, cleanupAfterInstall);
}

test.concurrent('install with arg that has install scripts', (): Promise<void> => {
  return runAdd({}, ['flow-bin'], 'install-with-arg-and-install-scripts');
});

test.concurrent('install with arg', (): Promise<void> => {
  return runAdd({}, ['is-online'], 'install-with-arg');
});

test.concurrent('install with --dev flag', (): Promise<void> => {
  return runAdd({dev: true}, ['left-pad@1.1.0'], 'add-with-flag', async (config) => {
    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));

    assert.deepEqual(pkg.devDependencies, {'left-pad': '1.1.0'});
    assert.deepEqual(pkg.dependencies, {});
  });
});

test.concurrent('install with --peer flag', (): Promise<void> => {
  return runAdd({peer: true}, ['left-pad@1.1.0'], 'add-with-flag', async (config) => {
    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));

    assert.deepEqual(pkg.peerDependencies, {'left-pad': '1.1.0'});
    assert.deepEqual(pkg.dependencies, {});
  });
});

test.concurrent('install with --optional flag', (): Promise<void> => {
  return runAdd({optional: true}, ['left-pad@1.1.0'], 'add-with-flag', async (config) => {
    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));

    assert.deepEqual(pkg.optionalDependencies, {'left-pad': '1.1.0'});
    assert.deepEqual(pkg.dependencies, {});
  });
});

test.concurrent('install with arg that has binaries', (): Promise<void> => {
  return runAdd({}, ['react-native-cli'], 'install-with-arg-and-bin');
});

test.concurrent('add should ignore cache', (): Promise<void> => {
  // left-pad@1.1.0 gets installed without --save
  // left-pad@1.1.0 gets installed with --save
  // files in mirror, yarn.lock, package.json and node_modules should reflect that

  return runAdd({}, ['left-pad@1.1.0'], 'install-save-to-mirror-when-cached', async (config, reporter) => {
    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '1.1.0',
    );

    const lockfile = await createLockfile(config.cwd);
    const install = new Add(['left-pad@1.1.0'], {}, config, reporter, lockfile);
    await install.init();
    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '1.1.0',
    );
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {'left-pad': '1.1.0'},
    );

    const mirror = await fs.walk(path.join(config.cwd, 'mirror-for-offline'));
    assert.equal(mirror.length, 1);
    assert.equal(mirror[0].relative, 'left-pad-1.1.0.tgz');

    const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines = explodeLockfile(lockFileWritten);
    assert.equal(lockFileLines[0], 'left-pad@1.1.0:');
    assert.equal(lockFileLines.length, 3);
    assert.notEqual(lockFileLines[2].indexOf('resolved left-pad-1.1.0.tgz'), -1);
  });
});

test.concurrent('add should not make package.json strict', (): Promise<void> => {
  return runAdd({}, ['left-pad@^1.1.0'], 'install-no-strict', async (config) => {
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {
        'left-pad': '^1.1.0',
        'mime-types': '^2.0.0',
      },
    );
  });
});

test.concurrent('add --save-exact should not make all package.json strict', (): Promise<void> => {
  return runAdd({saveExact: true}, ['left-pad@1.1.0'], 'install-no-strict-all', async (config) => {
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {
        'left-pad': '1.1.0',
        'mime-types': '^2.0.0',
      },
    );
  });
});

test.concurrent('add with new dependency should be deterministic 3', (): Promise<void> => {
  return runAdd({}, [], 'install-should-cleanup-when-package-json-changed-3', async (config, reporter) => {
    // expecting yarn check after installation not to fail

    await fs.copy(path.join(config.cwd, 'yarn.lock.after'), path.join(config.cwd, 'yarn.lock'));
    await fs.copy(path.join(config.cwd, 'package.json.after'), path.join(config.cwd, 'package.json'));

    const lockfile = await createLockfile(config.cwd);
    const install = new Install({}, config, reporter, lockfile);
    await install.init();

    let allCorrect = true;
    try {
      await check(config, reporter, {}, []);
    } catch (err) {
      allCorrect = false;
    }
    expect(allCorrect).toBe(true);
  });
});

test.concurrent('install --initMirror should add init mirror deps from package.json',
(): Promise<void> => {
  const mirrorPath = 'mirror-for-offline';
  const fixture = 'install-init-mirror';

  // initMirror gets converted to save flag in cli/install.js
  return runAdd({}, [], fixture, async (config) => {
    assert.equal(await getPackageVersion(config, 'mime-types'), '2.0.0');
    assert(semver.satisfies(await getPackageVersion(config, 'mime-db'), '~1.0.1'));

    const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
    assert.equal(mirror.length, 2);
    assert.equal(mirror[0].relative.indexOf('mime-db-1.0.'), 0);
    assert.equal(mirror[1].relative, 'mime-types-2.0.0.tgz');

    const lockFileContent = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines = explodeLockfile(lockFileContent);
    assert.equal(lockFileLines.length, 8);
    assert.equal(lockFileLines[0].indexOf('mime-db@'), 0);
    assert.equal(lockFileLines[3].indexOf('mime-types@2.0.0'), 0);
  });
});

test.concurrent('add with new dependency should be deterministic', (): Promise<void> => {
  // mime-types@2.0.0->mime-db@1.0.3 is saved in local mirror and is deduped
  // install mime-db@1.23.0 should move mime-db@1.0.3 deep into mime-types

  const mirrorPath = 'mirror-for-offline';
  const fixture = 'install-deterministic';

  return runInstall({}, path.join('..', 'add', fixture), async (config): Promise<void> => {
    assert(semver.satisfies(
      await getPackageVersion(config, 'mime-db'),
      '~1.0.1'),
    );
    assert.equal(
      await getPackageVersion(config, 'mime-types'),
      '2.0.0',
    );

    return runAdd({}, ['mime-db@1.23.0'], fixture, async (config) => {
      assert(semver.satisfies(
        await getPackageVersion(config, 'mime-db'),
        '1.23.0',
      ));
      assert.equal(
        await getPackageVersion(config, 'mime-types'),
        '2.0.0',
      );
      assert.equal(
        await getPackageVersion(config, 'mime-types/mime-db'),
        '1.0.3',
      );
      assert.deepEqual(
        JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies, {
          'mime-types': '2.0.0',
          'mime-db': '1.23.0',
        },
      );

      const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
      const lockFileLines = explodeLockfile(lockFileWritten);
      assert.equal(lockFileLines.length, 11);


      const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
      assert.equal(mirror.length, 3);
      assert.equal(mirror[1].relative, 'mime-db-1.23.0.tgz');
    });
  });
});

// TODO https://github.com/facebook/yarn/issues/79
xit('add with new dependency should be deterministic 2', (): Promise<void> => {
  // mime-types@2.0.0->mime-db@1.0.1 is saved in local mirror and is deduped
  // install mime-db@1.0.3 should replace mime-db@1.0.1 in root

  const mirrorPath = 'mirror-for-offline';
  const fixture = 'install-deterministic-2';

  return runInstall({}, path.join('..', 'add', fixture), async (config): Promise<void> => {
    assert.equal(
      await getPackageVersion(config, 'mime-db'),
      '1.0.1',
    );
    assert.equal(
      await getPackageVersion(config, 'mime-types'),
      '2.0.0',
    );

    return runAdd({}, ['mime-db@1.0.3'], fixture, async (config) => {
      assert.equal(
        await getPackageVersion(config, 'mime-db'),
        '1.0.3',
      );
      assert.equal(
        await getPackageVersion(config, 'mime-types'),
        '2.0.0',
      );
      assert(!await fs.exists(path.join(config.cwd, 'node_modules/mime-types/node-modules/mime-db')));
      assert.deepEqual(
        JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies, {
          'mime-types': '2.0.0',
          'mime-db': '1.0.3',
        },
      );

      const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
      const lockFileLines = explodeLockfile(lockFileWritten);
      assert.equal(lockFileLines.length, 8);

      const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
      assert.equal(mirror.length, 3);
      assert.equal(mirror[1].relative, 'mime-db-1.0.3.tgz');
    });
  });
});

test.concurrent('add with offline mirror', (): Promise<void> => {
  const mirrorPath = 'mirror-for-offline';
  return runAdd({}, ['is-array@^1.0.1'], 'install-with-save-offline-mirror', async (config) => {
    const allFiles = await fs.walk(config.cwd);

    assert(allFiles.findIndex((file): boolean => {
      return file.relative === path.join(mirrorPath, 'is-array-1.0.1.tgz');
    }) !== -1);

    const rawLockfile = await fs.readFile(path.join(config.cwd, constants.LOCKFILE_FILENAME));
    const lockfile = parse(rawLockfile);
    assert.equal(
      lockfile['is-array@^1.0.1']['resolved'],
      'is-array-1.0.1.tgz#e9850cc2cc860c3bc0977e84ccf0dd464584279a',
    );
  });
});

test.concurrent('install with --save and without offline mirror', (): Promise<void> => {
  const mirrorPath = 'mirror-for-offline';
  return runAdd({}, ['is-array@^1.0.1'], 'install-with-save-no-offline-mirror', async (config) => {

    const allFiles = await fs.walk(config.cwd);

    assert(allFiles.findIndex((file): boolean => {
      return file.relative === `${mirrorPath}/is-array-1.0.1.tgz`;
    }) === -1);

    const rawLockfile = await fs.readFile(path.join(config.cwd, constants.LOCKFILE_FILENAME));
    const lockfile = parse(rawLockfile);
    assert(
      lockfile['is-array@^1.0.1']['resolved']
        .indexOf('is-array-1.0.1.tgz#e9850cc2cc860c3bc0977e84ccf0dd464584279a') >= 0,
    );
  });
});

test.concurrent('upgrade scenario', (): Promise<void> => {
  // left-pad first installed 0.0.9 then updated to 1.1.0
  // files in mirror, yarn.lock, package.json and node_modules should reflect that

  const mirrorPath = 'mirror-for-offline';

  return runAdd({}, ['left-pad@0.0.9'], 'install-upgrade-scenario', async (config, reporter): Promise<void> => {
    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '0.0.9',
    );
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {'left-pad': '0.0.9'},
    );

    const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines = explodeLockfile(lockFileWritten);
    assert.equal(lockFileLines[0], 'left-pad@0.0.9:');
    assert.equal(lockFileLines.length, 3);
    assert.notEqual(lockFileLines[2].indexOf('resolved left-pad-0.0.9.tgz'), -1);

    const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
    assert.equal(mirror.length, 1);
    assert.equal(mirror[0].relative, 'left-pad-0.0.9.tgz');

    //
    const add = new Add(['left-pad@1.1.0'], {}, config, reporter, await Lockfile.fromDirectory(config.cwd));
    await add.init();

    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '1.1.0',
    );
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {'left-pad': '1.1.0'},
    );

    const lockFileWritten2 = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines2 = explodeLockfile(lockFileWritten2);
    assert.equal(lockFileLines2[0], 'left-pad@1.1.0:');
    assert.equal(lockFileLines2.length, 3);
    assert.notEqual(lockFileLines2[2].indexOf('resolved left-pad-1.1.0.tgz'), -1);

    const mirror2 = await fs.walk(path.join(config.cwd, mirrorPath));
    assert.equal(mirror2.length, 2);
    assert.equal(mirror2[1].relative, 'left-pad-1.1.0.tgz');
  });
});

test.concurrent('upgrade scenario 2 (with sub dependencies)', (): Promise<void> => {
  // mime-types@2.0.0 is saved in local mirror and gets updated to mime-types@2.1.11
  // files in mirror, yarn.lock, package.json and node_modules should reflect that

  const mirrorPath = 'mirror-for-offline';
  const fixture = 'install-upgrade-scenario-2';

  return runInstall({}, path.join('..', 'add', fixture), async (config): Promise<void> => {
    assert(semver.satisfies(
      await getPackageVersion(config, 'mime-db'),
      '~1.0.1'),
    );
    assert.equal(
      await getPackageVersion(config, 'mime-types'),
      '2.0.0',
    );

    return runAdd({}, ['mime-types@2.1.11'], fixture, async (config) => {
      assert(semver.satisfies(
        await getPackageVersion(config, 'mime-db'),
        '~1.23.0',
      ));
      assert.equal(
        await getPackageVersion(config, 'mime-types'),
        '2.1.11',
      );

      const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
      const lockFileLines = explodeLockfile(lockFileWritten);
      assert.equal(lockFileLines[0], 'mime-db@~1.23.0:');
      assert.notEqual(lockFileLines[2].indexOf('resolved mime-db-'), -1);
      assert.equal(lockFileLines[3], 'mime-types@2.1.11:');
      assert.notEqual(lockFileLines[5].indexOf('resolved mime-types-2.1.11.tgz'), -1);

      const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
      assert.equal(mirror.length, 4);
      const newFilesInMirror = mirror.filter((elem): boolean => {
        return elem.relative !== 'mime-db-1.0.3.tgz' && elem.relative !== 'mime-types-2.0.0.tgz';
      });

      assert.equal(newFilesInMirror.length, 2);
    });
  });
});

test.concurrent('downgrade scenario', (): Promise<void> => {
  // left-pad first installed 1.1.0 then downgraded to 0.0.9
  // files in mirror, yarn.lock, package.json and node_modules should reflect that

  return runAdd({}, ['left-pad@1.1.0'], 'install-downgrade-scenario', async (config, reporter): Promise<void> => {
    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '1.1.0',
    );
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {'left-pad': '1.1.0'},
    );

    const mirrorPath = 'mirror-for-offline';
    const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines = explodeLockfile(lockFileWritten);
    assert.equal(lockFileLines[0], 'left-pad@1.1.0:');
    assert.equal(lockFileLines.length, 3);
    assert.notEqual(lockFileLines[2].indexOf('resolved left-pad-1.1.0.tgz'), -1);

    const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
    assert.equal(mirror.length, 1);
    assert.equal(mirror[0].relative, 'left-pad-1.1.0.tgz');

    //

    const add = new Add(['left-pad@0.0.9'], {},  config, reporter, await Lockfile.fromDirectory(config.cwd));
    await add.init();

    assert.equal(
      await getPackageVersion(config, 'left-pad'),
      '0.0.9',
    );
    assert.deepEqual(
      JSON.parse(await fs.readFile(path.join(config.cwd, 'package.json'))).dependencies,
      {'left-pad': '0.0.9'},
    );

    const lockFileWritten2 = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines2 = explodeLockfile(lockFileWritten2);
    assert.equal(lockFileLines2[0], 'left-pad@0.0.9:');
    assert.equal(lockFileLines2.length, 3);
    assert.notEqual(lockFileLines2[2].indexOf('resolved left-pad-0.0.9.tgz'), -1);

    const mirror2 = await fs.walk(path.join(config.cwd, mirrorPath));
    assert.equal(mirror2.length, 2);
    assert.equal(mirror2[0].relative, 'left-pad-0.0.9.tgz');
  });
});

// https://github.com/yarnpkg/yarn/issues/318
test.concurrent('modules resolved multiple times should save to mirror correctly', (): Promise<void> => {
  // the package.json in this fixture has 4 transitive dependants on module which that should resolve to
  // which@^1.0.5, which@^1.1.1, which@^1.2.8, which@^1.2.9:
  //   version "1.2.11"
  //   resolved which-1.2.11.tgz#c8b2eeea6b8c1659fa7c1dd4fdaabe9533dc5e8b
  return runAdd({}, [], 'no-mirror-remote-when-duplicates', async (config): Promise<void> => {
    const mirrorPath = 'mirror-for-offline';

    // check that which module was downloaded to mirror
    const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
    const whichModule = mirror.find((elem): any => elem.relative.match(/which-1.*\.tgz/));
    expect(whichModule).toBeDefined();

    const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
    const lockFileLines = explodeLockfile(lockFileWritten);

    // no entry in lockfile that resolves to https
    const httpResolved = lockFileLines.find((elem): any => elem.match(/resolved "https:\/\//));
    expect(httpResolved).toBeUndefined();

    // which dependency must be resolved to file in local mirror
    const whichResolved = lockFileLines.find((elem): any => elem.match(/ {2}resolved which-1.*\.tgz#.*/));
    expect(whichResolved).toBeDefined();
  });
});

test.concurrent('add should put a git dependency to mirror', (): Promise<void> => {
  const mirrorPath = 'mirror-for-offline';

  return runAdd(
    {},
    ['mime-db@https://github.com/jshttp/mime-db.git#1.24.0'],
    'install-git-mirror',
    async (config, reporter): Promise<void> => {
      assert(semver.satisfies(
        await getPackageVersion(config, 'mime-db'),
        '1.24.0'),
      );
      const mirror = await fs.walk(path.join(config.cwd, mirrorPath));
      assert.equal(mirror.length, 1);
      expect(mirror[0].relative).toMatch(/mime-db\.git.*/);

      const lockFileWritten = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
      const lockFileLines = explodeLockfile(lockFileWritten);
      // lock file contains mirror resolved line
      expect(lockFileLines.find((line) => line.match(/.*resolved mime-db\.git\-.*/))).toBeDefined();

      // can reinstall, now from mirror
      await fs.unlink(path.join(config.cwd, 'node_modules'));

      //
      const install = new Install({}, config, reporter, await Lockfile.fromDirectory(config.cwd));
      await install.init();

      assert(semver.satisfies(
        await getPackageVersion(config, 'mime-db'),
        '1.24.0'),
      );
      await fs.unlink(path.join(config.cwd, mirrorPath));
      await fs.unlink(path.join(config.cwd, 'package.json'));
    },
  );
});
