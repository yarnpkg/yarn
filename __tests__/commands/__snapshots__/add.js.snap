exports[`test install with arg that has install scripts 1`] = `
"{
  \"dependencies\": {
    \"left-pad\": \"^1.1.0\",
    \"mime-types\": \"^2.0.0\"
  }
}
"
`;
